beacon_node/beacon_chain/src/beacon_fork_choice_store.rs

```diff
@@ -327,16 +331,16 @@ where
                 .map_err(Error::FailedToReadState)?
                 .ok_or_else(|| Error::MissingState(justified_block.state_root()))?;
 
-            self.justified_balances = JustifiedBalances::from_justified_state(&state)?;
+            let (justified_balances, slashed_indices) =
+                JustifiedBalances::from_justified_state_with_equivocating_indices(&state)?;
+            self.justified_balances = justified_balances;
+            self.equivocating_indices
+                .extend(slashed_indices.into_iter());
```

Comment:
We are missing similar logic in the cache hit case. When the cache is primed we don't update the equivocating indices:

https://github.com/sigp/lighthouse/blob/42a8f4feb9f497a1ddd2443ed9aafbc1cd523194/beacon_node/beacon_chain/src/beacon_fork_choice_store.rs#L113-L117

We could maybe store the slashed indices in the cache (schema migration 💀) or go slightly further off-spec and add them eagerly, even though the state isn't justified.

Reply:
As per a helpful suggestion from Potuz, I've opted to just set the balances for slashed validators to zero in `JustifiedBalances` in 5377ac612.

This change is not *perfect*, since it means that anything we cached pre-migration will have a non-zero balance for slashed validators whilst anything we compute without the cache will have a zero balance (i.e., cache inconsistency). I find this to be acceptable for a few reasons:

- The number of active-but-slashed-by-a-proposer-slashing validators is almost always zero.
- The cache discrepancy will only last until a newly-imported state is justified (1-2 epochs, assuming typical participation).
- This will only take effect for a short time on each node as they are updated, rather than happening on all nodes at once. Therefore it's likely that the portion of affected nodes on the network at any one time is likely to be small.

Reply:
After some consideration overnight and a chat with Michael and Sean, I've written up a DB migration to drop the balances cache to avoid inconsistencies at the expense of some DB reads at startup: [77a4970](https://github.com/sigp/lighthouse/pull/3962/commits/77a49708d7150f9808571b07017f9533e3c0ead0)